File: apps/web/src/lib/recommendation-outcomes.ts
import { validId } from './evaluation-history';
export const OUTCOMES_PATH = '/performance/recommendations';
export const FILTER_KEYS = ['symbol','from_date','to_date','horizon','benchmark','alpha'] as const;
export type Filters = Record<typeof FILTER_KEYS[number], string>;
export type OutcomeQuery = Filters & { before: string; through: string };
export type Measurement = {
 outcome_id:string; recommendation_id:string; thesis_id:string|null; symbol:string; currency_code:string;
 strategy_name:string; market_code:string; recommendation_date:string; recommendation_action:string; recommendation_score:number;
 measurement_start_date:string; measurement_end_date:string; horizon_days:number; nominal_horizon:number|null;
 entry_price:number; exit_price:number; absolute_return_pct:number; benchmark_code:string|null; benchmark_return_pct:number|null;
 alpha_pct:number|null; max_drawdown_pct:number|null; outcome_label:string; recorded_at:string; source_run_id:string|null;
 evaluation_snapshot:{eval_run_id:string;snapshot_id:string;as_of_date:string|null;recorded_at:string}|null;
};
export type OutcomeReport = {
 rows:Measurement[]; filters:Filters; generated_at:string; benchmarks:(string|null)[];
 summary:{measurement_count:number;recommendation_count:number;symbol_count:number;missing_alpha_count:number;first_recommendation_date:string|null;last_recommendation_date:string|null};
 pagination:{limit:number;returned_count:number;through:string;has_more:boolean;next_cursor:string|null};
};
export type OutcomeResult = { report:OutcomeReport|null; query:OutcomeQuery; issue:'invalid'|'unavailable'|null };
export const emptyQuery = ():OutcomeQuery => ({symbol:'',from_date:'',to_date:'',horizon:'',benchmark:'',alpha:'',before:'',through:''});
const record = (v:unknown):Record<string,unknown> => v !== null && typeof v==='object' && !Array.isArray(v) ? v as Record<string,unknown> : {};
const number = (v:unknown):v is number => typeof v==='number' && Number.isFinite(v);
const integer = (v:unknown):v is number => number(v) && Number.isSafeInteger(v) && v>=0;
export function dateOnly(v:unknown):v is string {
 if(typeof v!=='string'||!/^\d{4}-\d{2}-\d{2}$/.test(v)||v.startsWith('0000')) return false;
 const d=new Date(v+'T00:00:00Z');return Number.isFinite(d.getTime()) && d.toISOString().slice(0,10)===v;
}
const watermark = (v:unknown):v is string => v==='0'||validId(v);
export const cursorValid = (v:unknown):v is string => typeof v==='string' && v.split(':').length===2 && dateOnly(v.split(':')[0]) && validId(v.split(':')[1]);
export function parseOutcomeQuery(raw:Record<string,unknown>):OutcomeQuery {
 const query=emptyQuery();
 for(const key of Object.keys(raw)) {
  if(!Object.hasOwn(query,key)||typeof raw[key]!=='string') throw Error('invalid query');
  query[key as keyof OutcomeQuery]=raw[key] as string;
 }
 query.symbol=query.symbol.toUpperCase();
 if(query.symbol&&!/^[A-Z0-9][A-Z0-9._-]{0,19}$/.test(query.symbol)) throw Error('invalid symbol');
 if((query.from_date&&!dateOnly(query.from_date))||(query.to_date&&!dateOnly(query.to_date))||(query.from_date&&query.to_date&&query.from_date>query.to_date)) throw Error('invalid dates');
 if(!['','30','90','180','365','other'].includes(query.horizon)||!['','positive','negative','zero','missing'].includes(query.alpha)) throw Error('invalid filter');
 if(query.benchmark&&!/^[A-Za-z0-9_^][A-Za-z0-9 ^&._/-]{0,63}$/.test(query.benchmark)) throw Error('invalid benchmark');
 if((query.before&&!cursorValid(query.before))||(query.through&&!watermark(query.through))) throw Error('invalid cursor');
 return query;
}
export function outcomeQueryString(query:Partial<OutcomeQuery>):string {
 const p=new URLSearchParams();for(const key of Object.keys(emptyQuery()) as (keyof OutcomeQuery)[]) if(query[key]) p.set(key,query[key]!);
 return p.toString();
}
export const outcomeHref = (query:Partial<OutcomeQuery>) => `${OUTCOMES_PATH}${outcomeQueryString(query)?'?'+outcomeQueryString(query):''}`;
export function percent(value:number|null, alpha=false):string {
 return value===null?'미측정':`${value>0?'+':''}${(value*100).toLocaleString('ko-KR',{minimumFractionDigits:2,maximumFractionDigits:2})}${alpha?'%p':'%'}`;
}
export function evaluationHref(snapshot:NonNullable<Measurement['evaluation_snapshot']>):string {
 const after=BigInt(snapshot.snapshot_id)-1n;
 return `/performance/evaluations/${snapshot.eval_run_id}${after>0n?'?after='+after:''}`;
}
export function parseOutcomeReport(raw:unknown, query:OutcomeQuery):OutcomeReport {
 const r=record(raw), p=record(r.pagination), s=record(r.summary), filters=record(r.filters);
 if(r.contract_version!=='recommendation-outcome-explorer-v1'||r.source!=='live'||r.read_only!==true||r.horizon_tolerance_days!==7||r.record_basis!=='stored_measurements_with_current_recommendation_links') throw Error('invalid report');
 if(FILTER_KEYS.some(k=>filters[k]!==query[k])||typeof r.generated_at!=='string'||!Number.isFinite(Date.parse(r.generated_at))) throw Error('wrong scope');
 if(!watermark(p.through)||(query.through&&p.through!==query.through)||p.limit!==25||typeof p.has_more!=='boolean'||(p.next_cursor!==null&&!cursorValid(p.next_cursor))) throw Error('invalid pagination');
 if(p.has_more!==(p.next_cursor!==null)||!Array.isArray(r.rows)||r.rows.length>25||p.returned_count!==r.rows.length) throw Error('invalid page');
 if(!Array.isArray(r.benchmarks)||r.benchmarks.some(v=>v!==null&&typeof v!=='string')) throw Error('invalid benchmarks');
 for(const k of ['measurement_count','recommendation_count','symbol_count','missing_alpha_count']) if(!integer(s[k])) throw Error('invalid counts');
 if((s.measurement_count as number)<r.rows.length||(s.recommendation_count as number)>(s.measurement_count as number)||(s.symbol_count as number)>(s.recommendation_count as number)||(s.missing_alpha_count as number)>(s.measurement_count as number)) throw Error('invalid counts');
 for(const k of ['first_recommendation_date','last_recommendation_date']) if(s[k]!==null&&!dateOnly(s[k])) throw Error('invalid range');
 let previous=query.before?query.before.split(':'):null;
 const seen=new Set<string>();
 for(const value of r.rows) {
  const row=record(value);
  if(!validId(row.outcome_id)||!validId(row.recommendation_id)||(row.thesis_id!==null&&!validId(row.thesis_id))||(row.source_run_id!==null&&!validId(row.source_run_id))||seen.has(row.outcome_id)||BigInt(row.outcome_id)>BigInt(p.through)) throw Error('invalid identity');
  seen.add(row.outcome_id);
  for(const k of ['symbol','currency_code','strategy_name','market_code','recommendation_action','outcome_label','recorded_at']) if(typeof row[k]!=='string'||!row[k]) throw Error('missing context');
  if(!/^[A-Za-z0-9][A-Za-z0-9._-]{0,19}$/.test(row.symbol as string)||!dateOnly(row.recommendation_date)||!dateOnly(row.measurement_start_date)||!dateOnly(row.measurement_end_date)||row.measurement_start_date>row.measurement_end_date||!integer(row.horizon_days)) throw Error('invalid dates');
  for(const k of ['recommendation_score','entry_price','exit_price','absolute_return_pct']) if(!number(row[k])) throw Error('invalid metric');
  for(const k of ['benchmark_return_pct','alpha_pct','max_drawdown_pct']) if(row[k]!==null&&!number(row[k])) throw Error('invalid metric');
  if(row.benchmark_code!==null&&typeof row.benchmark_code!=='string') throw Error('invalid benchmark');
  if(row.nominal_horizon!==null&&![30,90,180,365].includes(row.nominal_horizon as number)) throw Error('invalid horizon');
  if(previous&&(row.measurement_end_date>previous[0]||(row.measurement_end_date===previous[0]&&BigInt(row.outcome_id)>=BigInt(previous[1])))) throw Error('invalid order');
  previous=[row.measurement_end_date,row.outcome_id];
  if(row.evaluation_snapshot!==null) {const e=record(row.evaluation_snapshot);if(!validId(e.eval_run_id)||!validId(e.snapshot_id)||(e.as_of_date!==null&&!dateOnly(e.as_of_date))||typeof e.recorded_at!=='string') throw Error('invalid snapshot');}
 }
 if(p.has_more&&(r.rows.length!==25||!previous||p.next_cursor!==previous.join(':'))) throw Error('invalid continuation');
 return r as unknown as OutcomeReport;
}
export async function loadRecommendationOutcomes(raw:Record<string,unknown>, options:{fetcher?:typeof fetch;timeoutMs?:number}={}):Promise<OutcomeResult> {
 let query:OutcomeQuery;try {query=parseOutcomeQuery(raw);} catch {return {report:null,query:emptyQuery(),issue:'invalid'};}
 const controller=new AbortController();let timer:ReturnType<typeof setTimeout>|undefined;
 try {
  const base=(process.env.STOCKANALYSIS_FRONTEND_API_BASE_URL??'http://127.0.0.1:8765').replace(/\/$/,'');
  const token=process.env.STOCKANALYSIS_FRONTEND_API_READ_TOKEN;
  const read=async()=>{
   const suffix=outcomeQueryString(query);
   const response=await(options.fetcher??fetch)(`${base}/api/recommendation-outcomes?limit=25${suffix?'&'+suffix:''}`,{method:'GET',cache:'no-store',redirect:'error',signal:controller.signal,headers:{Accept:'application/json',...(token?{Authorization:`Bearer ${token}`}:{})}});
   if(!response.ok) throw Error('unavailable');return parseOutcomeReport(await response.json(),query);
  };
  const deadline=new Promise<never>((_,reject)=>{timer=setTimeout(()=>{controller.abort();reject(Error('timeout'));},options.timeoutMs??8000);});
  return {report:await Promise.race([read(),deadline]),query,issue:null};
 } catch {return {report:null,query,issue:'unavailable'};}
 finally {clearTimeout(timer);}
}

File: src/stockanalysis/frontend/recommendation_outcomes.py
"""Read stored recommendation measurements independently of portfolio attribution."""
from __future__ import annotations

import json
import re
from dataclasses import asdict, dataclass
from datetime import date, datetime, timezone
from urllib.parse import parse_qsl, urlencode, urlsplit

from stockanalysis.frontend.recommendation_eval_history import EVAL_NAME, EvaluationHistoryError, MAX_ID
from stockanalysis.ingest.config import RuntimeConfig
from stockanalysis.ingest.macro.sql import sql_literal
from stockanalysis.ingest.psql import PsqlCommandExecutor
from stockanalysis.performance.outcome_window import outcome_window_match_sql

API_PATH = "/api/recommendation-outcomes"
VERSION = "recommendation-outcome-explorer-v1"
HORIZONS = (30, 90, 180, 365)


def invalid():
    return EvaluationHistoryError("Invalid recommendation outcome filters or cursor.", code="FrontendPaginationInvalid")


def unavailable():
    return EvaluationHistoryError("Stored recommendation outcomes are unavailable.", code="FrontendLiveReadUnavailable")


@dataclass(frozen=True)
class OutcomeRequest:
    symbol: str = ""
    from_date: str = ""
    to_date: str = ""
    horizon: str = ""
    benchmark: str = ""
    alpha: str = ""
    before: str = ""
    through: str = ""
    limit: int = 25

    def filters(self):
        return {k: v for k, v in asdict(self).items() if k not in {"before", "through", "limit"}}


def valid_date(value: str):
    if not re.fullmatch(r"[0-9]{4}-[0-9]{2}-[0-9]{2}", value):
        raise invalid()
    date.fromisoformat(value)


def valid_id(value: str, *, zero=False):
    if not re.fullmatch(r"0|[1-9][0-9]{0,18}" if zero else r"[1-9][0-9]{0,18}", value) or int(value) > MAX_ID:
        raise invalid()


def parse_outcome_request(path: str) -> OutcomeRequest:
    try:
        if any(ord(c) <= 32 or ord(c) == 127 for c in path):
            raise invalid()
        parsed = urlsplit(path)
        if parsed.path != API_PATH or parsed.scheme or parsed.netloc or parsed.fragment:
            raise invalid()
        values = {}
        for k, v in parse_qsl(parsed.query, keep_blank_values=True, strict_parsing=True, errors="strict", max_num_fields=9):
            if k not in OutcomeRequest.__dataclass_fields__ or k in values:
                raise invalid()
            values[k] = v
        symbol = values.get("symbol", "").upper()
        if symbol and not re.fullmatch(r"[A-Z0-9][A-Z0-9._-]{0,19}", symbol):
            raise invalid()
        values["symbol"] = symbol
        for key in ("from_date", "to_date"):
            if values.get(key):
                valid_date(values[key])
        if values.get("from_date") and values.get("to_date") and values["from_date"] > values["to_date"]:
            raise invalid()
        if values.get("horizon", "") not in {"", "other", *(str(h) for h in HORIZONS)}:
            raise invalid()
        if values.get("alpha", "") not in {"", "positive", "negative", "zero", "missing"}:
            raise invalid()
        benchmark = values.get("benchmark", "")
        if benchmark and not re.fullmatch(r"[A-Za-z0-9_^][A-Za-z0-9 ^&._/-]{0,63}", benchmark):
            raise invalid()
        if values.get("before"):
            day, identifier = values["before"].split(":")
            valid_date(day); valid_id(identifier)
        if values.get("through"):
            valid_id(values["through"], zero=True)
        if "limit" in values:
            valid_id(values["limit"])
            values["limit"] = int(values["limit"])
            if values["limit"] > 100:
                raise invalid()
        return OutcomeRequest(**values)
    except (ValueError, UnicodeError) as exc:
        raise invalid() from exc


def render_recommendation_outcomes_sql(request: OutcomeRequest) -> str:
    request = parse_outcome_request(API_PATH + "?" + urlencode(asdict(request)))
    conditions = []
    if request.symbol:
        conditions.append(f"upper(symbol) = {sql_literal(request.symbol)}")
    if request.from_date:
        conditions.append(f"recommendation_date >= {sql_literal(request.from_date)}::date")
    if request.to_date:
        conditions.append(f"recommendation_date <= {sql_literal(request.to_date)}::date")
    if request.horizon:
        conditions.append("nominal_horizon is null" if request.horizon == "other" else f"nominal_horizon = {int(request.horizon)}")
    if request.benchmark:
        conditions.append("benchmark_code is null" if request.benchmark == "_missing" else f"upper(benchmark_code) = upper({sql_literal(request.benchmark)})")
    if request.alpha:
        conditions.append({"positive": "alpha_pct > 0", "negative": "alpha_pct < 0", "zero": "alpha_pct = 0", "missing": "alpha_pct is null"}[request.alpha])
    where = " and ".join(conditions) or "true"
    ceiling = request.through or "coalesce((select max(outcome_id) from performance.recommendation_outcome), 0)"
    cursor = "true"
    if request.before:
        day, identifier = request.before.split(":")
        cursor = f"(measurement_end_date, outcome_id) < ({sql_literal(day)}::date, {int(identifier)}::bigint)"
    horizon_case = "case " + " ".join(
        "when " + outcome_window_match_sql(recommendation_id="r.recommendation_id", recommendation_date="b.as_of_date",
            horizon_days=str(h), end_date="o.measurement_end_date", alias="o") + f" then {h}"
        for h in HORIZONS) + " end"
    return f"""with ceiling as (select {ceiling}::bigint as id), population as (
    select o.*, b.as_of_date as recommendation_date, b.strategy_name, b.market_code,
        r.action as recommendation_action, r.total_score as recommendation_score,
        r.thesis_id, i.primary_symbol as symbol, i.currency_code, {horizon_case} as nominal_horizon
    from performance.recommendation_outcome o
    join signal.recommendation r using (recommendation_id)
    join signal.recommendation_batch b using (batch_id)
    join ref.instrument i using (instrument_id)
    where o.outcome_id <= (select id from ceiling)
), filtered as (select * from population where {where}), page as (
    select * from filtered where {cursor}
    order by measurement_end_date desc, outcome_id desc limit {request.limit + 1}
), page_rows as (
    select p.*, saved.snapshot
    from page p
    left join lateral (
        select json_build_object('eval_run_id',s.eval_run_id::text,'snapshot_id',s.snapshot_id::text,
            'as_of_date',e.as_of_date,'recorded_at',s.created_at) as snapshot
        from ai.recommendation_eval_snapshot s join ai.eval_run e using (eval_run_id)
        where s.source_outcome_id = p.outcome_id and s.source_recommendation_id = p.recommendation_id
            and e.eval_name = {sql_literal(EVAL_NAME)}
        order by s.created_at desc, s.snapshot_id desc limit 1
    ) saved on true
)
select json_build_object(
    'through', (select id::text from ceiling),
    'summary', (select json_build_object('measurement_count',count(*),'recommendation_count',count(distinct recommendation_id),
        'symbol_count',count(distinct symbol),'missing_alpha_count',count(*) filter(where alpha_pct is null),
        'first_recommendation_date',min(recommendation_date),'last_recommendation_date',max(recommendation_date)) from filtered),
    'benchmarks', coalesce((select json_agg(benchmark_code order by benchmark_code nulls last)
        from (select distinct benchmark_code from population) b),'[]'::json),
    'rows', coalesce((select json_agg(json_build_object(
        'outcome_id',outcome_id::text,'recommendation_id',recommendation_id::text,'thesis_id',thesis_id::text,
        'symbol',symbol,'currency_code',currency_code,'strategy_name',strategy_name,'market_code',market_code,
        'recommendation_date',recommendation_date,'recommendation_action',recommendation_action,'recommendation_score',recommendation_score,
        'measurement_start_date',measurement_start_date,'measurement_end_date',measurement_end_date,
        'horizon_days',horizon_days,'nominal_horizon',nominal_horizon,
        'entry_price',entry_price,'exit_price',exit_price,'absolute_return_pct',absolute_return_pct,
        'benchmark_code',benchmark_code,'benchmark_return_pct',benchmark_return_pct,'alpha_pct',alpha_pct,
        'max_drawdown_pct',max_drawdown_pct,'outcome_label',outcome_label,'recorded_at',created_at,
        'source_run_id',source_run_id::text,'evaluation_snapshot',snapshot
    ) order by measurement_end_date desc,outcome_id desc) from page_rows),'[]'::json)
)::text;"""


def resolve_recommendation_outcomes(path: str, *, source: str, config=None, executor=None):
    request = parse_outcome_request(path)
    if source not in {"live", "auto"}:
        raise unavailable()
    try:
        if executor is None:
            executor = PsqlCommandExecutor.from_config(config or RuntimeConfig.from_env())
        raw = json.loads(executor.execute_scalar(render_recommendation_outcomes_sql(request)))
        valid_id(raw["through"], zero=True)
        rows = raw["rows"]
        if not isinstance(rows, list) or len(rows) > request.limit + 1:
            raise unavailable()
        keys = []
        for row in rows:
            valid_id(row["outcome_id"]); valid_id(row["recommendation_id"])
            valid_date(row["measurement_end_date"])
            keys.append((row["measurement_end_date"], int(row["outcome_id"])))
            if int(row["outcome_id"]) > int(raw["through"]):
                raise unavailable()
        if keys != sorted(set(keys), reverse=True):
            raise unavailable()
        if request.before:
            day, identifier = request.before.split(":")
            if any(key >= (day, int(identifier)) for key in keys):
                raise unavailable()
        summary = raw["summary"]
        for field in ("measurement_count", "recommendation_count", "symbol_count", "missing_alpha_count"):
            if type(summary[field]) is not int or summary[field] < 0:
                raise unavailable()
        if summary["measurement_count"] < len(rows):
            raise unavailable()
        shown = rows[:request.limit]
        next_cursor = f"{shown[-1]['measurement_end_date']}:{shown[-1]['outcome_id']}" if len(rows) > request.limit else None
        return {"contract_version": VERSION, "source": "live", "read_only": True,
                "generated_at": datetime.now(timezone.utc).isoformat(), "filters": request.filters(),
                "summary": summary, "benchmarks": raw["benchmarks"], "rows": shown,
                "pagination": {"limit": request.limit, "returned_count": len(shown), "through": raw["through"],
                               "has_more": next_cursor is not None, "next_cursor": next_cursor},
                "record_basis": "stored_measurements_with_current_recommendation_links",
                "horizon_tolerance_days": 7}
    except Exception as exc:
        # Do not turn failed reads into zero outcomes, or expose SQL/credentials.
        raise unavailable() from exc
